testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/providers/sse/SseEventSinkClosingTest.java

```diff
@@ -0,0 +1,108 @@
+import org.junit.runner.RunWith;
+
+@RunWith(Arquillian.class)
+@RunAsClient
+public class SseEventSinkClosingTest {
+
+	private final static Logger logger = Logger.getLogger(SseEventSinkClosingTest.class);
+
+	@Deployment
+	public static Archive<?> deploy() {
+		WebArchive war = TestUtil.prepareArchive(SseEventSinkClosingTest.class.getSimpleName());
+		war.addClass(SseEventSinkClosingTestResource.class);
+		war.addClass(SseEventSinkClosingTestResource.ContainerFilter.class);
+		return TestUtil.finishContainerPrepare(war, null, Arrays.asList(SseEventSinkClosingTestResource.class),
+				SseEventSinkClosingTestResource.ContainerFilter.class);
+	}
+
+	private String generateURL() {
+		return PortProviderUtil.generateBaseUrl(SseEventSinkClosingTest.class.getSimpleName());
+	}
+
+	@After
+	public void reset() throws Exception {
+		Client client = ClientBuilder.newClient();
+		try {
+			client.target(generateURL()).path(SseEventSinkClosingTestResource.BASE_PATH)
+					.path(SseEventSinkClosingTestResource.RESET_RESPONSE_FILTER_INVOCATION_COUNT_PATH).request()
+					.delete();
+		} finally {
+			client.close();
+		}
+	}
+
+	@Test
+	public void Should_NotInvokeResponseFilter_When_ClosingSseEventSinkOnceMessageSent() throws Exception {
+
+		Client client = ClientBuilder.newClient();
+		try {
+
+			WebTarget baseTarget = client.target(generateURL()).path(SseEventSinkClosingTestResource.BASE_PATH);
+
+			try (Response response = baseTarget.path(SseEventSinkClosingTestResource.SEND_AND_CLOSE_PATH)
+					.request(MediaType.SERVER_SENT_EVENTS_TYPE).get()) {
+				Assert.assertEquals(Status.OK.getStatusCode(), response.getStatus());
+			}
+
+			try (Response response = baseTarget
+					.path(SseEventSinkClosingTestResource.GET_RESPONSE_FILTER_INVOCATION_COUNT_PATH)
+					.request(MediaType.TEXT_PLAIN_TYPE).get()) {
+				Assert.assertEquals(Status.OK.getStatusCode(), response.getStatus());
+				Assert.assertEquals(Integer.valueOf(1), response.readEntity(Integer.class));
+			}
+
+		} finally {
+			client.close();
+		}
+
+	}
+
+	@Test
+	public void Should_InvokeResponseFilterOnlyOnce_When_ClosingSseEventSinkWithoutSending() throws Exception {
+
+		Client client = ClientBuilder.newClient();
+		try {
+
+			WebTarget baseTarget = client.target(generateURL()).path(SseEventSinkClosingTestResource.BASE_PATH);
+
+			try (Response response = baseTarget.path(SseEventSinkClosingTestResource.CLOSE_WITHOUT_SENDING_PATH)
+					.request(MediaType.SERVER_SENT_EVENTS_TYPE).get()) {
+				Assert.assertEquals(Status.NO_CONTENT.getStatusCode(), response.getStatus());
+			}
+
+			try (Response response = baseTarget
+					.path(SseEventSinkClosingTestResource.GET_RESPONSE_FILTER_INVOCATION_COUNT_PATH)
+					.request(MediaType.TEXT_PLAIN_TYPE).get()) {
+				Assert.assertEquals(Status.OK.getStatusCode(), response.getStatus());
+				Assert.assertEquals(Integer.valueOf(1), response.readEntity(Integer.class));
+			}
+
+		} finally {
+			client.close();
+		}
+
+	}
+
+}
```

Comment:
@NicoNes  SseEventSink is internally constructed with a servlet response.  Should it all invoke response filter when close SseEventSink, no matter the sse event is sent or not ?

Reply:
Based on my understanding of the spec **9.5 Processing Pipeline**, the filters must only been called when sending the initial response, the one with the headers:

> For compatibility purposes, implementations MUST initiate processing of an SSE response when either the
> first message is sent or when the resource method returns, whichever happens first. The initial SSE response,
> which may only include the HTTP headers, is processed using the standard JAX-RS pipeline as described
> in Appendix C. Each subsequent SSE event may include a different payload and thus require the use of a
> specific message body writer. Note that since this use case differs slightly from the normal JAX-RS pipeline,
> implementations SHOULD NOT call entity interceptors on each individual event 1.

So when the first 200 response is sent, either when resource method returns or when a sse event is sent whichever happens first the filters must be called and the headers will be committed.
Then if closing the SseEventSink after that, the filters should not be called and by the way  sending any new response header such as 204 will be useless since headers has already been commited, it will never be received by client.
So to be coherent with the spirit of the specs if headers has already been committed it is uselless to called filters again which can be confusing since headers will never be received by client.

However when closing the SseEventSink if no sse event has been sent, the 204 response should be sent to the client as the initial response and the filters must be called

Reply:
Just to understand the change for this issue.The cleanable is called after resource method returns, so closedWithoutSending() will call SseEventOuputImpl#close() without process pipeline. The pipeline is processed at these lines  ?

> 	 if(this.closed)
> 	 {
> 		 jaxrsResponse = (BuiltResponse) Response.noContent().build();
> 	 }

Reply:
Yes calling SseEventOuputImpl#close() will just complete the asynch response without processing the pipeline.
The processing will be done at line 101 (https://github.com/NicoNes/Resteasy/blob/2edd2389102cbfc199d881de0dadff73707b3071/resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/providers/sse/SseEventOutputImpl.java#L101)
>ServerResponseWriter.writeNomapResponse(jaxrsResponse, request, response,
> ResteasyProviderFactory.getInstance(), t -> {
>}, true);

Reply:
Thanks @NicoNes . Do you think these are better test names in SseEventSinkClosingTest?
~~Should_NotInvokeResponseFilter_When_ClosingSseEventSinkOnceMessageSent ()~~  to testFilterForEventSent() 
~~Should_InvokeResponseFilterOnlyOnce_When_ClosingSseEventSinkWithoutSending()~~ to testFilterForMethodReturn()

Reply:
Personaly I think that the convention I used is clearer. It allows a quick and better understanding of what is being tested , under what cicumstances and what is the expected result whithout having to read the test code. (https://dzone.com/articles/7-popular-unit-test-naming)

But it's personnal and if it's not the resteasy convention let me know and I will update the names.

Reply:
We now name most of resteasy tests with the name convention #3 or something similar in the linked doc : 
>test[Feature being tested]: This one makes it easy to read the test as the feature to be tested is written as part of test name. Although, there are arguments that the “test” prefix is redundant. However, some sections of developer love to use this technique

Reply:
OK names switched to the one you proposed.